File: backend_db/SDKs/typescript/src/realtime.ts
/**
 * SelfDB SDK Realtime Module
 * 
 * Phoenix Channels WebSocket implementation for realtime updates.
 */

import { HttpClient } from './client';
import { RealtimePayload, RealtimeCallback, RealtimeEvent } from './models';

/**
 * WebSocket interface that works in both Node.js and browser
 */
interface WebSocketLike {
    readonly readyState: number;
    onopen: ((event: unknown) => void) | null;
    onclose: ((event: unknown) => void) | null;
    onerror: ((event: unknown) => void) | null;
    onmessage: ((event: { data: string }) => void) | null;
    send(data: string): void;
    close(): void;
}

/**
 * Get WebSocket constructor (works in both Node.js and browser)
 */
async function getWebSocketClass(): Promise<new (url: string) => WebSocketLike> {
    if (typeof WebSocket !== 'undefined') {
        return WebSocket as unknown as new (url: string) => WebSocketLike;
    }
    // Node.js - dynamically import ws
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const ws = await import(/* webpackIgnore: true */ 'ws');
    return (ws.default || ws) as unknown as new (url: string) => WebSocketLike;
}

/**
 * Phoenix message format
 */
interface PhoenixMessage {
    topic: string;
    event: string;
    payload: Record<string, unknown>;
    ref: string | null;
}

/**
 * Channel state
 */
type ChannelState = 'closed' | 'joining' | 'joined' | 'leaving';

/**
 * Realtime channel for subscribing to events
 */
export class RealtimeChannel {
    private state: ChannelState = 'closed';
    private handlers: Map<string, RealtimeCallback[]> = new Map();
    private joinRef: string | null = null;

    constructor(
        public readonly topic: string,
        private realtime: Realtime
    ) {}

    /**
     * Register an event handler
     */
    on(event: RealtimeEvent | '*', callback: RealtimeCallback): RealtimeChannel {
        const eventKey = event === '*' ? '*' : event;
        const handlers = this.handlers.get(eventKey) || [];
        handlers.push(callback);
        this.handlers.set(eventKey, handlers);
        return this;
    }

    /**
     * Remove an event handler
     */
    off(event: RealtimeEvent | '*', callback?: RealtimeCallback): RealtimeChannel {
        const eventKey = event === '*' ? '*' : event;
        if (callback) {
            const handlers = this.handlers.get(eventKey) || [];
            const index = handlers.indexOf(callback);
            if (index > -1) {
                handlers.splice(index, 1);
            }
            this.handlers.set(eventKey, handlers);
        } else {
            this.handlers.delete(eventKey);
        }
        return this;
    }

    /**
     * Subscribe to the channel (sends phx_join)
     */
    async subscribe(): Promise<void> {
        if (this.state === 'joined' || this.state === 'joining') {
            return;
        }

        this.state = 'joining';
        this.joinRef = this.realtime.generateRef();

        await this.realtime.send({
            topic: this.topic,
            event: 'phx_join',
            payload: {},
            ref: this.joinRef,
        });

        this.state = 'joined';
    }

    /**
     * Unsubscribe from the channel (sends phx_leave)
     */
    async unsubscribe(): Promise<void> {
        if (this.state === 'closed' || this.state === 'leaving') {
            return;
        }

        this.state = 'leaving';

        await this.realtime.send({
            topic: this.topic,
            event: 'phx_leave',
            payload: {},
            ref: this.realtime.generateRef(),
        });

        this.state = 'closed';
        this.handlers.clear();
    }

    /**
     * Handle incoming message for this channel
     */
    handleMessage(event: string, payload: Record<string, unknown>): void {
        // Create realtime payload
        const realtimePayload: RealtimePayload = {
            event: event as RealtimeEvent,
            table: (payload.table as string) || this.topic.replace('table:', ''),
            new: (payload.new as Record<string, unknown>) || null,
            old: (payload.old as Record<string, unknown>) || null,
            raw: payload,
        };

        // Call specific event handlers
        const handlers = this.handlers.get(event) || [];
        for (const handler of handlers) {
            try {
                handler(realtimePayload);
            } catch (error) {
                console.error(`Error in realtime handler for ${event}:`, error);
            }
        }

        // Call wildcard handlers
        const wildcardHandlers = this.handlers.get('*') || [];
        for (const handler of wildcardHandlers) {
            try {
                handler(realtimePayload);
            } catch (error) {
                console.error('Error in realtime wildcard handler:', error);
            }
        }
    }

    /**
     * Get channel state
     */
    getState(): ChannelState {
        return this.state;
    }
}

/**
 * Realtime connection state
 */
type ConnectionState = 'disconnected' | 'connecting' | 'connected' | 'disconnecting';

/**
 * Realtime module for Phoenix Channels WebSocket
 */
export class Realtime {
    private socket: WebSocketLike | null = null;
    private state: ConnectionState = 'disconnected';
    private channels: Map<string, RealtimeChannel> = new Map();
    private refCounter = 0;
    private heartbeatInterval: ReturnType<typeof setInterval> | null = null;
    private pendingMessages: PhoenixMessage[] = [];
    private messageResolvers: Map<string, { resolve: () => void; reject: (error: Error) => void }> = new Map();

    constructor(private client: HttpClient, private baseUrl: string, private apiKey: string) {}

    /**
     * Generate a unique reference for messages
     */
    generateRef(): string {
        this.refCounter++;
        return String(this.refCounter);
    }

    /**
     * Get WebSocket URL
     */
    private getWebSocketUrl(): string {
        const wsBaseUrl = this.baseUrl.replace(/^http/, 'ws');
        const token = this.client.getAccessToken() || '';
        return `${wsBaseUrl}/realtime/socket?X-API-Key=${this.apiKey}&token=${token}`;
    }

    /**
     * Connect to the realtime server
     */
    async connect(): Promise<void> {
        if (this.state === 'connected' || this.state === 'connecting') {
            return;
        }

        this.state = 'connecting';

        // Get WebSocket class (browser or Node.js)
        const WebSocketClass = await getWebSocketClass();
        const url = this.getWebSocketUrl();

        return new Promise((resolve, reject) => {
            const socket = new WebSocketClass(url);
            this.socket = socket;

            socket.onopen = () => {
                this.state = 'connected';
                this.startHeartbeat();
                
                // Send any pending messages
                for (const message of this.pendingMessages) {
                    this.sendImmediate(message);
                }
                this.pendingMessages = [];
                
                resolve();
            };

            socket.onclose = () => {
                this.state = 'disconnected';
                this.stopHeartbeat();
            };

            socket.onerror = () => {
                if (this.state === 'connecting') {
                    reject(new Error('Failed to connect to realtime server'));
                }
            };

            socket.onmessage = (event) => {
                this.handleMessage(event.data);
            };
        });
    }

    /**
     * Disconnect from the realtime server
     */
    async disconnect(): Promise<void> {
        if (this.state === 'disconnected' || this.state === 'disconnecting') {
            return;
        }

        this.state = 'disconnecting';
        this.stopHeartbeat();

        // Unsubscribe from all channels
        for (const channel of this.channels.values()) {
            await channel.unsubscribe();
        }
        this.channels.clear();

        if (this.socket) {
            this.socket.close();
            this.socket = null;
        }

        this.state = 'disconnected';
    }

    /**
     * Create or get a channel for a topic
     */
    channel(topic: string): RealtimeChannel {
        let channel = this.channels.get(topic);
        if (!channel) {
            channel = new RealtimeChannel(topic, this);
            this.channels.set(topic, channel);
        }
        return channel;
    }

    /**
     * Send a message
     */
    async send(message: PhoenixMessage): Promise<void> {
        if (this.state !== 'connected' || !this.socket) {
            this.pendingMessages.push(message);
            return;
        }

        return new Promise((resolve, reject) => {
            if (message.ref) {
                this.messageResolvers.set(message.ref, { resolve, reject });
                
                // Timeout for response
                setTimeout(() => {
                    if (this.messageResolvers.has(message.ref!)) {
                        this.messageResolvers.delete(message.ref!);
                        resolve(); // Don't reject, just resolve (best effort)
                    }
                }, 5000);
            }

            this.sendImmediate(message);
            
            if (!message.ref) {
                resolve();
            }
        });
    }

    /**
     * Send a message immediately
     */
    private sendImmediate(message: PhoenixMessage): void {
        if (this.socket && this.socket.readyState === 1) { // 1 = OPEN
            // Phoenix protocol uses array format: [join_ref, ref, topic, event, payload]
            const phoenixMessage = [null, message.ref, message.topic, message.event, message.payload];
            this.socket.send(JSON.stringify(phoenixMessage));
        }
    }

    /**
     * Handle incoming WebSocket message
     */
    private handleMessage(data: string): void {
        try {
            // Phoenix protocol uses array format: [join_ref, ref, topic, event, payload]
            const [_joinRef, ref, topic, event, payload] = JSON.parse(data);

            // Handle reply messages
            if (event === 'phx_reply' && ref && this.messageResolvers.has(ref)) {
                const resolver = this.messageResolvers.get(ref)!;
                this.messageResolvers.delete(ref);
                
                if (payload.status === 'ok') {
                    resolver.resolve();
                } else {
                    resolver.reject(new Error(payload.response?.reason || 'Unknown error'));
                }
                return;
            }

            // Route to appropriate channel
            const channel = this.channels.get(topic);
            if (channel) {
                channel.handleMessage(event, payload || {});
            }
        } catch (error) {
            console.error('Error handling realtime message:', error);
        }
    }

    /**
     * Start heartbeat interval
     */
    private startHeartbeat(): void {
        this.heartbeatInterval = setInterval(() => {
            this.send({
                topic: 'phoenix',
                event: 'heartbeat',
                payload: {},
                ref: this.generateRef(),
            });
        }, 30000); // 30 seconds
    }

    /**
     * Stop heartbeat interval
     */
    private stopHeartbeat(): void {
        if (this.heartbeatInterval) {
            clearInterval(this.heartbeatInterval);
            this.heartbeatInterval = null;
        }
    }

    /**
     * Get connection state
     */
    getState(): ConnectionState {
        return this.state;
    }
}

File: backend_db/SDKs/swift/Sources/SelfDB/Realtime.swift
import Foundation

/// Realtime event types
public enum RealtimeEvent: String, Sendable {
    case insert = "INSERT"
    case update = "UPDATE"
    case delete = "DELETE"
    case all = "*"
}

/// Payload received in realtime callbacks
public struct RealtimePayload: @unchecked Sendable {
    public let event: RealtimeEvent
    public let table: String
    public let new: [String: Any]?
    public let old: [String: Any]?
    public let raw: [String: Any]
    
    init(event: RealtimeEvent, table: String, new: [String: Any]?, old: [String: Any]?, raw: [String: Any]) {
        self.event = event
        self.table = table
        self.new = new
        self.old = old
        self.raw = raw
    }
}

/// Callback type for realtime events
public typealias RealtimeCallback = @Sendable (RealtimePayload) -> Void

/// Realtime channel for subscribing to events
public final class RealtimeChannel: @unchecked Sendable {
    private let topic: String
    private weak var realtime: Realtime?
    private var eventHandlers: [RealtimeEvent: [RealtimeCallback]] = [:]
    private let lock = NSLock()
    
    init(topic: String, realtime: Realtime) {
        self.topic = topic
        self.realtime = realtime
    }
    
    /// Register a handler for a specific event type
    @discardableResult
    public func on(_ event: RealtimeEvent, handler: @escaping RealtimeCallback) -> RealtimeChannel {
        lock.lock()
        defer { lock.unlock() }
        
        if eventHandlers[event] == nil {
            eventHandlers[event] = []
        }
        eventHandlers[event]?.append(handler)
        return self
    }
    
    /// Subscribe to the channel
    public func subscribe() async throws {
        try await realtime?.joinChannel(topic: topic)
    }
    
    /// Unsubscribe from the channel
    public func unsubscribe() async throws {
        try await realtime?.leaveChannel(topic: topic)
    }
    
    /// Handle an incoming event
    func handleEvent(_ eventType: RealtimeEvent, payload: RealtimePayload) {
        lock.lock()
        let handlers = eventHandlers[eventType] ?? []
        let allHandlers = eventHandlers[.all] ?? []
        lock.unlock()
        
        for handler in handlers {
            handler(payload)
        }
        for handler in allHandlers {
            handler(payload)
        }
    }
    
    /// Get the topic name
    func getTopic() -> String {
        return topic
    }
}

/// Realtime module for WebSocket-based real-time updates using Phoenix Channels
public final class Realtime: @unchecked Sendable {
    private let baseUrl: String
    private let apiKey: String
    private weak var client: HTTPClient?
    
    private var webSocket: URLSessionWebSocketTask?
    private var session: URLSession?
    private var channels: [String: RealtimeChannel] = [:]
    private var messageRef: Int = 0
    private var heartbeatTask: Task<Void, Never>?
    private var receiveTask: Task<Void, Never>?
    private var _isConnected: Bool = false
    private let lock = NSLock()
    
    /// Whether the realtime connection is active
    public var isConnected: Bool {
        lock.lock()
        defer { lock.unlock() }
        return _isConnected
    }
    
    init(baseUrl: String, apiKey: String, client: HTTPClient) {
        self.baseUrl = baseUrl
        self.apiKey = apiKey
        self.client = client
    }
    
    /// Connect to the realtime server
    public func connect() async throws {
        lock.lock()
        guard !_isConnected else {
            lock.unlock()
            return
        }
        lock.unlock()
        
        // Get the access token
        guard let token = await client?.getAccessToken() else {
            throw SelfDBError.authenticationError
        }
        
        // Build WebSocket URL (matching Python/TypeScript SDKs)
        var wsUrl = baseUrl
            .replacingOccurrences(of: "http://", with: "ws://")
            .replacingOccurrences(of: "https://", with: "wss://")
        
        if wsUrl.hasSuffix("/") {
            wsUrl = String(wsUrl.dropLast())
        }
        wsUrl += "/realtime/socket?X-API-Key=\(apiKey)&token=\(token)"
        
        guard let url = URL(string: wsUrl) else {
            throw SelfDBError.badRequest("Invalid WebSocket URL")
        }
        
        session = URLSession(configuration: .default)
        webSocket = session?.webSocketTask(with: url)
        webSocket?.resume()
        
        lock.lock()
        _isConnected = true
        lock.unlock()
        
        // Start receiving messages
        startReceiving()
        
        // Start heartbeat
        startHeartbeat()
    }
    
    /// Disconnect from the realtime server
    public func disconnect() async throws {
        lock.lock()
        _isConnected = false
        heartbeatTask?.cancel()
        heartbeatTask = nil
        receiveTask?.cancel()
        receiveTask = nil
        
        let channelsCopy = channels
        channels.removeAll()
        lock.unlock()
        
        // Leave all channels
        for (topic, _) in channelsCopy {
            try? await leaveChannel(topic: topic)
        }
        
        webSocket?.cancel(with: .goingAway, reason: nil)
        webSocket = nil
        session = nil
    }
    
    /// Create a channel for a topic
    public func channel(_ topic: String) -> RealtimeChannel {
        lock.lock()
        defer { lock.unlock() }
        
        if let existing = channels[topic] {
            return existing
        }
        
        let channel = RealtimeChannel(topic: topic, realtime: self)
        channels[topic] = channel
        return channel
    }
    
    /// Join a channel (internal)
    func joinChannel(topic: String) async throws {
        guard isConnected else {
            throw SelfDBError.webSocketError("Not connected")
        }
        
        let ref = nextRef()
        try await sendPhoenixMessage(joinRef: nil, ref: ref, topic: topic, event: "phx_join", payload: [:])
    }
    
    /// Leave a channel (internal)
    func leaveChannel(topic: String) async throws {
        guard isConnected else { return }
        
        let ref = nextRef()
        try await sendPhoenixMessage(joinRef: nil, ref: ref, topic: topic, event: "phx_leave", payload: [:])
        
        lock.lock()
        channels.removeValue(forKey: topic)
        lock.unlock()
    }
    
    /// Get next message reference
    private func nextRef() -> String {
        lock.lock()
        defer { lock.unlock() }
        messageRef += 1
        return String(messageRef)
    }
    
    /// Send a Phoenix message in array format: [join_ref, ref, topic, event, payload]
    private func sendPhoenixMessage(joinRef: String?, ref: String, topic: String, event: String, payload: [String: Any]) async throws {
        guard let webSocket = webSocket else {
            throw SelfDBError.webSocketError("WebSocket not connected")
        }
        
        // Phoenix protocol uses array format
        let message: [Any] = [joinRef as Any, ref, topic, event, payload]
        let data = try JSONSerialization.data(withJSONObject: message)
        let string = String(data: data, encoding: .utf8) ?? "[]"
        
        try await webSocket.send(.string(string))
    }
    
    /// Start receiving messages
    private func startReceiving() {
        receiveTask = Task { [weak self] in
            while let self = self, self.isConnected {
                do {
                    guard let webSocket = self.webSocket else { break }
                    let message = try await webSocket.receive()
                    
                    switch message {
                    case .string(let text):
                        self.handleMessage(text)
                    case .data(let data):
                        if let text = String(data: data, encoding: .utf8) {
                            self.handleMessage(text)
                        }
                    @unknown default:
                        break
                    }
                } catch {
                    if self.isConnected {
                        // Connection error
                        break
                    }
                }
            }
        }
    }
    
    /// Handle incoming message in Phoenix array format: [join_ref, ref, topic, event, payload]
    private func handleMessage(_ text: String) {
        guard let data = text.data(using: .utf8),
              let array = try? JSONSerialization.jsonObject(with: data) as? [Any],
              array.count >= 5 else { return }
        
        // Phoenix array format: [join_ref, ref, topic, event, payload]
        let topic = array[2] as? String ?? ""
        let event = array[3] as? String ?? ""
        let payload = array[4] as? [String: Any] ?? [:]
        
        // Handle Phoenix system events
        if event == "phx_reply" || event == "phx_close" {
            return
        }
        
        // Handle broadcast events (INSERT, UPDATE, DELETE)
        // Note: Phoenix may send lowercase ('insert', 'update', 'delete')
        let normalizedEvent = event.uppercased()
        if normalizedEvent == "INSERT" || normalizedEvent == "UPDATE" || normalizedEvent == "DELETE" {
            let eventType: RealtimeEvent
            switch normalizedEvent {
            case "INSERT": eventType = .insert
            case "UPDATE": eventType = .update
            case "DELETE": eventType = .delete
            default: return
            }
            
            // Extract table name from topic (format: "table:name")
            let tableName = topic.hasPrefix("table:") ? String(topic.dropFirst(6)) : topic
            
            // Build payload
            let newData = payload["new"] as? [String: Any]
            let oldData = payload["old"] as? [String: Any]
            
            let realtimePayload = RealtimePayload(
                event: eventType,
                table: tableName,
                new: newData,
                old: oldData,
                raw: payload
            )
            
            // Dispatch to channel
            lock.lock()
            let channel = channels[topic]
            lock.unlock()
            
            channel?.handleEvent(eventType, payload: realtimePayload)
        }
    }
    
    /// Start heartbeat
    private func startHeartbeat() {
        heartbeatTask = Task { [weak self] in
            while let self = self, self.isConnected {
                try? await Task.sleep(nanoseconds: 30_000_000_000) // 30 seconds
                
                guard self.isConnected else { break }
                
                let ref = self.nextRef()
                try? await self.sendPhoenixMessage(joinRef: nil, ref: ref, topic: "phoenix", event: "heartbeat", payload: [:])
            }
        }
    }
}
